File: hang/src/broadcast.rs
use crate::track::TrackConsumer;
use crate::{Audio, Catalog, CatalogConsumer, CatalogProducer, Result, TrackProducer, Video};
use moq_lite::Track;
use web_async::spawn;

pub use moq_lite::Broadcast;

#[derive(Clone)]
pub struct BroadcastProducer {
	pub catalog: CatalogProducer,
	pub tracks: moq_lite::BroadcastMap,
}

impl BroadcastProducer {
	pub fn new(producer: moq_lite::BroadcastProducer) -> Self {
		let producer = producer.map();

		let catalog = Catalog::default().produce();
		producer.insert(catalog.consume().track);

		Self {
			tracks: producer,
			catalog,
		}
	}

	pub fn consume(&self) -> BroadcastConsumer {
		self.tracks.inner.consume().into()
	}

	pub fn path(&self) -> &str {
		&self.tracks.inner.info.path
	}

	/// Add a video track to the broadcast.
	pub fn add_video(&mut self, track: TrackConsumer, info: Video) {
		self.tracks.insert(track.inner.clone());
		self.catalog.add_video(info.clone());
		self.catalog.publish();

		let mut this = self.clone();
		spawn(async move {
			let _ = track.closed().await;
			this.tracks.remove(&track.inner.info.name);
			this.catalog.remove_video(&info);
			this.catalog.publish();
		});
	}

	/// Add an audio track to the broadcast.
	pub fn add_audio(&mut self, track: TrackConsumer, info: Audio) {
		self.tracks.insert(track.inner.clone());
		self.catalog.add_audio(info.clone());
		self.catalog.publish();

		let mut this = self.clone();
		spawn(async move {
			let _ = track.closed().await;
			this.tracks.remove(&track.inner.info.name);
			this.catalog.remove_audio(&info);
			this.catalog.publish();
		});
	}

	pub fn create_video(&mut self, video: Video) -> TrackProducer {
		let producer: TrackProducer = video.track.clone().produce().into();
		self.add_video(producer.consume(), video);
		producer
	}

	pub fn create_audio(&mut self, audio: Audio) -> TrackProducer {
		let producer: TrackProducer = audio.track.clone().produce().into();
		self.add_audio(producer.consume(), audio);
		producer
	}
}

impl From<moq_lite::BroadcastProducer> for BroadcastProducer {
	fn from(inner: moq_lite::BroadcastProducer) -> Self {
		Self::new(inner)
	}
}

#[derive(Clone)]
pub struct BroadcastConsumer {
	pub inner: moq_lite::BroadcastConsumer,
}

impl BroadcastConsumer {
	pub fn new(inner: moq_lite::BroadcastConsumer) -> Self {
		Self { inner }
	}

	pub async fn catalog(&self) -> Result<CatalogConsumer> {
		let track = Track {
			name: Catalog::DEFAULT_NAME.to_string(),
			priority: -1,
		};
		Ok(self.inner.request(track).await?.into())
	}

	/// Subscribes to a track
	pub async fn track(&self, track: Track) -> Result<TrackConsumer> {
		let track = self.inner.request(track).await?;
		Ok(TrackConsumer::new(track))
	}
}

impl From<moq_lite::BroadcastConsumer> for BroadcastConsumer {
	fn from(inner: moq_lite::BroadcastConsumer) -> Self {
		Self::new(inner)
	}
}

File: js/moq/src/broadcast.ts
import { TrackConsumer, TrackProducer } from "./track";
import { WatchConsumer, WatchProducer } from "./util/watch";

class State {
	tracks = new Map<string, TrackConsumer>();

	// Called when a track is not found.
	// The receiver is responsible for producing the track.
	onUnknown?: (track: TrackProducer) => void;
}

/**
 * Handles writing and managing tracks in a broadcast.
 *
 * @public
 */
export class BroadcastProducer {
	/** The path identifier for this broadcast */
	#state: WatchProducer<State>;

	/**
	 * Creates a new BroadcastProducer with the specified path and state.
	 * @param state - The state producer
	 *
	 * @internal
	 */
	constructor() {
		this.#state = new WatchProducer<State>(new State());
	}

	/**
	 * Creates a new track with the specified name.
	 * @param name - The name of the track to create
	 * @returns A TrackProducer for the new track
	 */
	createTrack(name: string): TrackProducer {
		const track = new TrackProducer(name, 0);
		this.insertTrack(track.consume());
		return track;
	}

	/**
	 * Inserts an existing track into the broadcast.
	 * @param track - The track reader to insert
	 */
	insertTrack(track: TrackConsumer) {
		this.#state.update((state) => {
			const existing = state.tracks.get(track.name);
			existing?.close();
			state.tracks.set(track.name, track);
			return state;
		});
	}

	/**
	 * Removes a track from the broadcast.
	 * @param name - The name of the track to remove
	 */
	removeTrack(name: string) {
		try {
			this.#state.update((state) => {
				const track = state.tracks.get(name);
				track?.close();
				state.tracks.delete(name);
				return state;
			});
		} catch {
			// We don't care if we try to remove when closed
		}
	}

	/**
	 * Sets a callback for handling unknown (on-demand) tracks.
	 * If not specified, unknown tracks will be closed with a "not found" error.
	 *
	 * @param fn - The callback function to handle unknown tracks
	 */
	unknownTrack(fn?: (track: TrackProducer) => void) {
		this.#state.update((state) => {
			state.onUnknown = fn;
			return state;
		});
	}

	/**
	 * Aborts the writer with an error.
	 * @param reason - The error reason for aborting
	 */
	abort(reason: Error) {
		this.#state.abort(reason);
	}

	/**
	 * Closes the writer and all associated tracks.
	 */
	close() {
		this.#state.update((state) => {
			for (const track of state.tracks.values()) {
				track.close();
			}
			state.tracks.clear();
			return state;
		});
		this.#state.close();
	}

	/**
	 * Returns a promise that resolves when the writer is unused.
	 */
	async unused(): Promise<void> {
		return this.#state.unused();
	}

	consume(): BroadcastConsumer {
		return new BroadcastConsumer(this.#state.consume());
	}
}

/**
 * Handles reading and subscribing to tracks in a broadcast.
 *
 * @remarks `clone()` can be used to create multiple consumers, just remember to `close()` them.
 *
 * @public
 */
export class BroadcastConsumer {
	#state: WatchConsumer<State>;

	/**
	 * Creates a new BroadcastConsumer with the specified path and state.
	 * @param path - The path identifier
	 * @param state - The state consumer
	 *
	 * @internal
	 */
	constructor(state: WatchConsumer<State>) {
		this.#state = state;
	}

	/**
	 * Subscribes to a track with the specified priority.
	 * @param track - The name of the track to subscribe to
	 * @param priority - The priority level for the subscription
	 * @returns A TrackConsumer for the subscribed track
	 */
	subscribe(track: string, priority: number): TrackConsumer {
		const state = this.#state.value();

		const existing = state.tracks.get(track);
		if (existing) {
			return existing.clone();
		}

		const producer = new TrackProducer(track, priority);
		const consumer = producer.consume();

		if (state.onUnknown) {
			state.onUnknown(producer);
		} else {
			producer.abort(new Error("not found"));
		}

		return consumer;
	}

	/**
	 * Returns a promise that resolves when the reader is closed.
	 * @returns A promise that resolves when closed
	 */
	async closed(): Promise<void> {
		return this.#state.closed();
	}

	/**
	 * Closes the reader.
	 */
	close() {
		this.#state.close();
	}

	/**
	 * Creates a new instance of the reader using the same state.
	 * @returns A new BroadcastConsumer instance
	 */
	clone(): BroadcastConsumer {
		return new BroadcastConsumer(this.#state.clone());
	}
}
